oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtAuthenticationManagerResolver.java

```diff
@@ -0,0 +1,159 @@
+
+	/**
+	 * Construct a {@link JwtAuthenticationManagerResolver} using the provided parameters
+	 *
+	 * @param trustedIssuers a whitelist of trusted issuers
+	 */
+	public JwtAuthenticationManagerResolver(String... trustedIssuers) {
+		this(Arrays.asList(trustedIssuers));
+	}
+
+	/**
+	 * Construct a {@link JwtAuthenticationManagerResolver} using the provided parameters
+	 *
+	 * @param trustedIssuers a whitelist of trusted issuers
+	 */
+	public JwtAuthenticationManagerResolver(Collection<String> trustedIssuers) {
+		Assert.notEmpty(trustedIssuers, "trustedIssuers cannot be empty");
+		trustedIssuers = Collections.unmodifiableCollection(trustedIssuers);
+		this.issuerAuthenticationManagerResolver =
+				new TrustedIssuerJwtAuthenticationManagerResolver(trustedIssuers::contains);
+	}
+
+	/**
+	 * Construct a {@link JwtAuthenticationManagerResolver} using the provided parameters
+	 *
+	 * @param issuerAuthenticationManagerResolver a strategy for resolving the {@link AuthenticationManager} by the issuer
+	 */
+	public JwtAuthenticationManagerResolver(AuthenticationManagerResolver<String> issuerAuthenticationManagerResolver) {
+		Assert.notNull(issuerAuthenticationManagerResolver, "issuerAuthenticationManagerResolver cannot be null");
+		this.issuerAuthenticationManagerResolver = issuerAuthenticationManagerResolver;
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public AuthenticationManager resolve(HttpServletRequest request) {
+		String issuer = this.issuerConverter.convert(request);
+		AuthenticationManager authenticationManager = this.issuerAuthenticationManagerResolver.resolve(issuer);
+		if (authenticationManager == null) {
+			throw new OAuth2AuthenticationException(invalidToken("Invalid issuer " + issuer));
+		}
+		return authenticationManager;
+	}
+
+	private static class JwtClaimIssuerConverter
+			implements Converter<HttpServletRequest, String> {
+
+		private final BearerTokenResolver resolver = new DefaultBearerTokenResolver();
+
+		@Override
+		public String convert(@NonNull HttpServletRequest request) {
+			String token = this.resolver.resolve(request);
+			try {
+				return JWTParser.parse(token).getJWTClaimsSet().getIssuer();
+			} catch (Exception e) {
+				throw new OAuth2AuthenticationException(invalidToken(e.getMessage()));
+			}
+		}
+	}
+
+	private static class TrustedIssuerJwtAuthenticationManagerResolver
+			implements AuthenticationManagerResolver<String> {
+
+		private final Map<String, AuthenticationManager> authenticationManagers = new ConcurrentHashMap<>();
+		private final Predicate<String> trustedIssuer;
+
+		TrustedIssuerJwtAuthenticationManagerResolver(Predicate<String> trustedIssuer) {
+			this.trustedIssuer = trustedIssuer;
+		}
+
+		@Override
+		public AuthenticationManager resolve(String issuer) {
+			if (this.trustedIssuer.test(issuer)) {
+				return this.authenticationManagers.computeIfAbsent(issuer, k -> {
+					JwtDecoder jwtDecoder = JwtDecoders.fromIssuerLocation(issuer);
```

Comment:
Do we need to document that changes on the authorization server will require a restart, since we only resolve metadata on the first request. Or should this be controlled? How do we do it when we don't use a resolver?

Reply:
I think that we are okay to leave this out since this is the way that it behaves like when there is no resolver.